packages/mdc-shape/foundation.js

```diff
@@ -0,0 +1,419 @@
+    } else {
+      const factor = 1 / numbers.ELEVATION_ANIMATION_FRAME_COUNT;
+      this.ambientShadowBlurIncrement_ =
+              (this.finalAmbientShadowBlur_ - this.ambientShadowBlur_) * factor;
+      this.ambientShadowOffsetYIncrement_ =
+              (this.finalAmbientShadowOffsetY_ - this.ambientShadowOffsetY_) * factor;
+      this.penumbraShadowBlurIncrement_ =
+              (this.finalPenumbraShadowBlur_ - this.penumbraShadowBlur_) * factor;
+      this.penumbraShadowOffsetYIncrement_ =
+              (this.finalPenumbraShadowOffsetY_ - this.penumbraShadowOffsetY_) * factor;
+      this.umbraShadowBlurIncrement_ =
+              (this.finalUmbraShadowBlur_ - this.umbraShadowBlur_) * factor;
+      this.umbraShadowOffsetYIncrement_ =
+              (this.finalUmbraShadowOffsetY_ - this.umbraShadowOffsetY_) * factor;
+      this.umbraShadowSpreadIncrement_ =
+              (this.finalUmbraShadowSpread_ - this.umbraShadowSpread_) * factor;
+      this.elevating_ = true;
+      this.animationFrameId_ = requestAnimationFrame(() => this.elevate_());
+    }
+  }
+
+  redraw() {
+    this.clear_();
+    this.init();
+    this.draw_();
+  }
+
+  generateClipPath_() {
+    return this.generatePath_(
+      this.adapter_.getCanvasWidth() - (numbers.SHADOW_PADDING * 2),
+      this.adapter_.getCanvasHeight() - (numbers.SHADOW_PADDING * 2),
+      0);
+  }
+
+  generateShadowPath_(shadowSpread) {
+    return this.generatePath_(
+      this.adapter_.getCanvasWidth(),
+      this.adapter_.getCanvasHeight(),
+      numbers.SHADOW_PADDING - shadowSpread);
+  }
+
+  generatePath_( /* width: number, height: number, padding: number */) {
+    // Subclasses should override this method to generate path data given a width, height, and padding
+    return 'm 0 0';
+  }
+
+  elevate_() {
+    this.ambientShadowBlur_ += this.ambientShadowBlurIncrement_;
+    this.ambientShadowOffsetY_ += this.ambientShadowOffsetYIncrement_;
+    this.penumbraShadowBlur_ += this.penumbraShadowBlurIncrement_;
+    this.penumbraShadowOffsetY_ += this.penumbraShadowOffsetYIncrement_;
+    this.umbraShadowBlur_ += this.umbraShadowBlurIncrement_;
+    this.umbraShadowOffsetY_ += this.umbraShadowOffsetYIncrement_;
+    this.umbraShadowSpread_ += this.umbraShadowSpreadIncrement_;
+
+    let finished = false;
+    if (Math.abs(this.finalAmbientShadowBlur_ - this.ambientShadowBlur_)
+          <= Math.abs(this.ambientShadowBlurIncrement_)) {
+      this.ambientShadowBlur_ = this.finalAmbientShadowBlur_;
+      this.ambientShadowOffsetY_ = this.finalAmbientShadowOffsetY_;
+      this.penumbraShadowBlur_ = this.finalPenumbraShadowBlur_;
+      this.penumbraShadowOffsetY_ = this.finalPenumbraShadowOffsetY_;
+      this.umbraShadowBlur_ = this.finalUmbraShadowBlur_;
+      this.umbraShadowOffsetY_ = this.finalUmbraShadowOffsetY_;
+      this.umbraShadowSpread_ = this.finalUmbraShadowSpread_;
+      finished = true;
+    }
+
+    this.redraw();
```

Comment:
Seems like stuff is happening in `init()` that doesn't need to happen every time `redraw()` is called. Doesn't seem destructive yet, just seems unnecessary.

Reply:
I moved some of the canvas stuff to a separate method. This way we don't re-create2dRenderingContext on every call to redraw.

The sync with canvas is important though, because the canvas could have changed size.